cmd/release-controller/sync_release_payload.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/release-controller/pkg/apis/release/v1alpha1"
+	releasecontroller "github.com/openshift/release-controller/pkg/release-controller"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+	"sort"
+)
+
+func (c *Controller) processReleasePayloadEvent(obj interface{}) {
+	switch t := obj.(type) {
+	case *v1alpha1.ReleasePayload:
+		klog.Infof("Processing ReleasePayload Event for: %s/%s", t.Namespace, t.Name)
+		c.addReleasePayloadQueueKey(queueKey{namespace: t.Namespace, name: t.Name})
+	default:
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %T", obj))
+	}
+}
+
+func (c *Controller) syncReleasePayload(key queueKey) error {
+	// Get the ReleasePayload resource with this namespace/name
+	payload, err := c.releasePayloadLister.ReleasePayloads(key.namespace).Get(key.name)
+	if err != nil {
+		// The ReleasePayload resource may no longer exist, in which case we stop processing.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("payload '%s/%s' in work queue no longer exists", key.namespace, key.name))
+			return nil
+		}
+		return err
+	}
+	klog.Infof("Syncing ReleasePayload: '%s/%s'", payload.Namespace, payload.Name)
+
+	// This is the entry point, for the release-controller, to respond to any updates to ReleasePayloads as they occur...
+
+	// TODO: Add business logic here
+
+	return nil
+}
+
+func (c *Controller) ensureReleasePayload(release *releasecontroller.Release, tagName string) (*v1alpha1.ReleasePayload, error) {
+	payload, err := c.releasePayloadLister.ReleasePayloads(release.Target.Namespace).Get(tagName)
+	if err == nil {
+		klog.V(4).Infof("ReleasePayload: %s/%s already exists", payload.Namespace, payload.Name)
+		return payload, nil
+	}
+	if !errors.IsNotFound(err) {
+		return nil, err
+	}
+	payload = newReleasePayload(release, tagName, c.jobNamespace, c.prowNamespace)
+	payload, err = c.releasePayloadClient.ReleasePayloads(release.Target.Namespace).Create(context.TODO(), payload, metav1.CreateOptions{})
+	if err == nil {
+		klog.V(4).Infof("ReleasePayload: %s/%s created", payload.Namespace, payload.Name)
+		return payload, nil
+	}
+	if !errors.IsAlreadyExists(err) {
+		return nil, err
+	}
+	// perform a live lookup if we are racing to create the payload
```

Comment:
if you have an error creating this, doesn't it mean that another thread is calling `ensureReleasePayload`, so we can simply fail the thread that made the concurrent call?

Reply:
Updated logic